src/typecheck.rs

```diff
@@ -1,11 +1,1384 @@
+            // must be numeric types with bounds
+            match (&inner_type, typ) {
+                (Type::NumT(_), Type::NumT(_)) => {
+                    // The cast is valid, return the target type
+                    Ok(typ.clone())
+                }
+                (_, Type::NumT(_)) => {
+                    // The inner expression is not a numeric type
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: inner_type,
+                    })
+                }
+                (Type::NumT(_), _) => {
+                    // The target type is not a numeric type
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: typ.clone(),
+                    })
+                }
+                _ => {
+                    // Neither the inner expression nor the target type are numeric types
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: inner_type,
+                    })
+                }
+            }
+        }
+        // Call expressions are always of the type of the function
+        Expr::Call { callee, args } => {
+            // Type check each argument
+            let mut arg_types: Vec<Type> = Vec::new();
+            for arg in args {
+                arg_types.push(typeck_expr(arg, ctx)?);
+            }
+
+            // Look up the function in the function environment
+            let function = ctx
+                .funcs
+                .get(callee)
+                .ok_or(TypeError::UndefinedFunction(*callee))?;
+
+            // Check that the number of arguments matches the number of parameters
+            if arg_types.len() != function.params.len() {
+                return Err(TypeError::WrongNumberOfArguments {
+                    expected: function.params.len(),
+                    actual: arg_types.len(),
+                });
+            }
+
+            // Verify that argument types match parameter types
+            for (arg_type, (_, param_type)) in arg_types.iter().zip(function.params.iter()) {
+                if arg_type != param_type {
+                    return Err(TypeError::TypeMismatch {
+                        expected: param_type.clone(),
+                        actual: arg_type.clone(),
+                    });
+                }
+            }
+
+            // Return the function's return type
+            Ok(function.ret_typ.clone())
+        }
+        // Pattern matching is always of the type of the scrutinee
+        Expr::Match { scrutinee, cases } => {
+            // Type check the scrutinee
+            let scrutinee_type = typeck_expr(scrutinee, ctx)?;
+
+            // If there are no cases, return an error
+            if cases.is_empty() {
```

Comment:
this isn't an error if you are doing proper exhaustivity check, if you aren't doing proper exhaustivity check then you aren't ensuring that `match` doesn't get stuck anyway; so I am OK with removing this because _the typing rules don't mandate this check._ However, you can enforce this in a way that makes the `result_type` stuff below nicer (see my comment below).